### DISCUSSION TITLE
What's the best approach on deployment

### DISCUSSION THREAD

**Discussion by vaisoliviu** [codefori/vscode-ibmi#1956] [Q&A] [answered]
Hello guys,

I have some question related to the deployment and build directly on the IFS.

I am currently using Codefori for some time, and I love it.
Found some of Liam's videos where he explains the use of Git and deployment+build on the IFS, and decided to give it a try.

I am currently developing on the AS400 machine, on an IASP. In order to have access to the IASP in VSCode I modified the JOBD for my user to set this IASP as the default. I have the sources in a number of libraries, depending on the client, but they are all same IASP.

I would love to be able to write the code locally and deploy it and compile it on the IFS, but it doesn't work for now for me 😞.
I copied all the sources to a local folder, and set up the deployment to the IFS as shown by Liam, but when I try to compile i get some errors.
Then I was thinking, what if the program that I am compiling needs a file or another object in order to be able to compile?
I did not create a make file, i'm not sure if I need one or not, I remember seeing a tutorial where it worked without it ???

Could u give me a hint to the best solution i could use?

**Comment by worksofliam** [comment]
@vaisoliviu Are you able to share some of those errors you are seeing? This will help us solve them!

**Reply by vaisoliviu** [reply]
Hello,

Here is the IBM i Deployment output:

<< Deployment started using method "changed"

Created deployment tarball C:\Users\vaiso\AppData\Local\Temp\tmp-26192-iUVwx4vu3SIx-.tar
Uploaded deployment tarball as /tmp/deploy_O_dbKgeADt.tar

/tmp/deploy_O_dbKgeADt.tar deleted
C:\Users\vaiso\AppData\Local\Temp\tmp-26192-iUVwx4vu3SIx-.tar deleted
Deployment failed: Error: Tarball extraction failed: .vscode: Error 3014 occurred.
tar: can't create .vscode/actions.json: No such file or directory  >>

**Reply by worksofliam** [reply]
@vaisoliviu What version of Code for IBM i are you using?

cc @sebjulliand

**Reply by vaisoliviu** [reply]
Version 2.8.0

**Reply by sebjulliand** [reply]
Hi @vaisoliviu ,
try adding a `.gitignore` file at the root of your workspace with this content:
```
.*
!.gitignore
```

Then try to deploy again.

**Reply by vaisoliviu** [reply]
Ok, this time the deployment worked.

My .gitignore looks like this:
![image](https://github.com/codefori/vscode-ibmi/assets/51542033/a0473a20-0e0a-421c-81da-f3e4042ff80e)

Now I get another error:
<< Commands:
                CRTBNDRPG PGM(COMP/rbekobl) SRCSTMF('CLIENT1/QRPGSRC/rbekobl.rpgle') OPTION(*EVENTF) DBGVIEW(*SOURCE) TGTCCSID(*JOB)

CPE3025:  No such path or directory.
RNS9339:  Impossible d'ouvrir le fichier CLIENT1/QRPGSRC/rbekobl.rpgle.

>>

**Reply by sebjulliand** [reply]
Just `.*` and `!.gitignore` should have worked since `.*` ignore everything that starts with `.`.
Anyway, now, the current directory doesn't seem to be set correctly. Can you share the entire `Code for i` output? There may something in there that may help.

**Reply by vaisoliviu** [reply]
So, here's everything:

<< Executing task: IBM i: Create RPGLE Program 

Running Action: Create RPGLE Program (7:07:05 PM)
Current library: COMP
Library list: CLIENT2 QTEMP IAFTP FAXN2OBJ FACDUPNEW FGE50LYL CLIENT1 CLIENT3 CLIENT4 MAP400V7 DEV CLIENT5 COMP
Working directory: /QSYS.LIB/COMP.LIB/
Commands:
                CRTBNDRPG PGM(COMP/rbekobl) SRCSTMF('CLIENT1/QRPGSRC/rbekobl.rpgle') OPTION(*EVENTF) DBGVIEW(*SOURCE) TGTCCSID(*JOB)

CPE3025:  No such path or directory.
RNS9339:  Impossible d'ouvrir le fichier CLIENT1/QRPGSRC/rbekobl.rpgle.
CPC2206:  Propriété de l'objet RETURNCODE de QTEMP, type *DTAARA modifiée.
CPC0904:  Zone de données RETURNCODE créée dans bibliothèque QTEMP.
RNS9310:  La compilation a échoué. Le programme RBEKOBL n'a pas été créé dans la bibliothèque COMP.
CPC2206:  Propriété de l'objet QZSHSYSTEM de QTEMP, type *USRSPC modifiée.

Fetching errors for COMP/RBEKOBL. >>

Maybe some of the libraries do not exist on the SYSBAS

**Reply by sebjulliand** [reply]
How...did you get a QSYS library (i.e. `/QSYS.LIB/COMP.LIB`) as a working directory?!
This should be an IFS folder, specifically your deployment folder.

Try this: from the IFS browser, go to your home directory, create a `build` folder, then right click on it and select `Set Deploy Workspace Location`:
![image](https://github.com/codefori/vscode-ibmi/assets/11096890/cda5a391-d88c-4b46-ad4c-37816faf21d8)

Then try running your action again.

**Reply by vaisoliviu** [reply]
I thought that i can use the library as a working directory, this way it would have been easier to copy the program in the destination library.

I did create the folder as u sayd, but the actions has the same result:

> Running Action: Create RPGLE Program (8:00:05 AM)
> Current library: COMP
> Library list: CLIENT2 QTEMP IAFTP FAXN2OBJ FACDUPNEW FGE50LYL CLIENT1 CLIENT3 CLIENT4 MAP400V7 DEV CLIENT5 COMP
> Working directory: /home/build/
> Commands:
 >               CRTBNDRPG PGM(COMP/rbekobl) SRCSTMF('CLIENT1/QRPGSRC/rbekobl.rpgle') OPTION(*EVENTF) DBGVIEW(*SOURCE) 
 >TGTCCSID(*JOB)

> CPE3025:  No such path or directory.
> RNS9339:  Impossible d'ouvrir le fichier CLIENT1/QRPGSRC/rbekobl.rpgle.
> CPC2206:  Propriété de l'objet RETURNCODE de QTEMP, type *DTAARA modifiée.
> CPC0904:  Zone de données RETURNCODE créée dans bibliothèque QTEMP.
> RNS9310:  La compilation a échoué. Le programme RBEKOBL n'a pas été créé dans la bibliothèque COMP.
> CPC2206:  Propriété de l'objet QZSHSYSTEM de QTEMP, type *USRSPC modifiée.

> Fetching errors for COMP/RBEKOBL.

 > *  The terminal process terminated with exit code: 1.

Also tried changing the library to QGPL, same results.

**Reply by sebjulliand** [reply, marked as answer]
By default, the actions are defined to compile programs into `&CURLIB` (the current library, as defined in the `User Library List` view). Like this:
```
CRTBNDRPG PGM(&CURLIB/&NAME) SRCSTMF('&RELATIVEPATH') OPTION(*EVENTF) DBGVIEW(*SOURCE) TGTCCSID(*JOB)
```
That is one way to define where the program is going to be created. But you may want to manage this compilation library differently. You can use custom variables.
You can either define a custom variable using the dedicated action from the palette:
![image](https://github.com/codefori/vscode-ibmi/assets/11096890/3a0a5c8a-9ef4-4f55-8312-4c9fa717e321)
These variables are stored in your user settings and linked to your connection settings, so they are active when connected to the IBM i they've been defined for. Using this, you could define a `&BUILDLIB` variable:
![image](https://github.com/codefori/vscode-ibmi/assets/11096890/a716e126-c942-41c6-8112-dbc97e6c1a10)

And then use it in your action definition:
```
CRTBNDRPG PGM(&BUILDLIB/&NAME) SRCSTMF('&RELATIVEPATH') OPTION(*EVENTF) DBGVIEW(*SOURCE) TGTCCSID(*JOB)
```

But since you're working with a workspace and local files, an even better way to handle variable for the project is the `.env` file. It has to be created at the root of the workspace. You can define in it as many variables as you want (key=value pairs, one per line):
![image](https://github.com/codefori/vscode-ibmi/assets/11096890/d00e0b43-4213-457e-88cd-708b19b4eca3)

Actions will pick up variables from the `.env` file when executed.

This being said, I'd advise that you move the build directory you created to your own home directory, not the `/home` directory itself. That is the root of every user's home directory and it should not contain a build directory. Your home directory is `/home/<your profile>`. Move your build directory there and make it your new deployment folder. Then run the action again and paste the output if it still fails.

**Reply by vaisoliviu** [reply]
Yep, it sure works now. U are freaking awesome, love the work and help that u do. 
Can I support your work somehow?

One small question, the program was placed in the QGPL library although I changed the actions.json to:
![image](https://github.com/codefori/vscode-ibmi/assets/51542033/4551c172-3662-495a-9c4a-5510c56fd4f3)

and created the .dev file as u instructed me:
![image](https://github.com/codefori/vscode-ibmi/assets/51542033/f5f66def-e470-42f9-846f-b036d2829711)

So the program is placed in the library that is declared as current, did a test with COMP here and it's places there:
![image](https://github.com/codefori/vscode-ibmi/assets/51542033/111e9354-1c83-4619-be25-39f68ccd1450)

**Reply by sebjulliand** [reply]
You're very welcome! Glad it worked 😄 
I would not expect the program to still be created in `&CURLIB` since you changed your action to use `&BUILDLIB`. Was the command in the output consistent with your action command pattern?

**Reply by vaisoliviu** [reply]
The command in the output is made with QGPL:
![image](https://github.com/codefori/vscode-ibmi/assets/51542033/649f2427-3c45-438c-9b20-19951e46f65f)

**Reply by vaisoliviu** [reply]
I created a variable and it works this way.
![image](https://github.com/codefori/vscode-ibmi/assets/51542033/ab61c881-d12d-4c54-8c5a-be16df0fdfff)

THANK YOU 🙏

**Reply by sebjulliand** [reply]
Perfect!
